src/coprocessor/trie/mod.rs

```diff
@@ -0,0 +1,734 @@
+            hash_cache,
+        );
+
+        let new_verified = self.new_proof.verify(new_root, key, new_value, hash_cache);
+
+        let paths_differ_by_at_most_one = self
+            .old_proof
+            .preimage_path
+            .iter()
+            .zip(&self.new_proof.preimage_path)
+            .all(|(a, b)| {
+                let differing_postition_count = a
+                    .iter()
+                    .zip(b)
+                    .map(|(x, y)| x != y)
+                    .fold(0, |acc, are_diff| acc + are_diff as usize);
+                differing_postition_count == 1 || differing_postition_count == 0
+            });
+
+        old_verified && new_verified && paths_differ_by_at_most_one
+    }
+}
+
+impl<'a, F: LurkField, const ARITY: usize, const HEIGHT: usize> Trie<'a, F, ARITY, HEIGHT> {
+    fn compute_hash(hash_cache: &PoseidonCache<F>, preimage: [F; ARITY]) -> F {
+        macro_rules! hash {
+            ($hash_name:ident, $n:expr) => {{
+                let mut buffer = [F::zero(); $n];
+                buffer.copy_from_slice(&preimage);
+
+                hash_cache.$hash_name(&buffer)
+            }};
+        }
+        match ARITY {
+            3 => hash!(hash3, 3),
+            4 => hash!(hash4, 4),
+            6 => hash!(hash6, 6),
+            8 => hash!(hash8, 8),
+            _ => unimplemented!(),
+        }
+    }
+
+    pub fn root(&self) -> F {
+        self.root
+    }
+
+    /// How many leaves does this `Trie` have?
+    pub const fn leaves(&self) -> usize {
+        self.row_size(HEIGHT)
+    }
+
+    /// How many nodes does the `row`th row have?
+    pub const fn row_size(&self, row: usize) -> usize {
+        debug_assert!(row <= HEIGHT);
+        (ARITY as u32).pow(row as u32) as usize
+    }
+
+    fn register_hash(&mut self, preimage: [F; ARITY]) -> F {
+        let hash = Self::compute_hash(self.hash_cache, preimage);
+
+        self.children.insert(FWrap(hash), preimage);
+        hash
+    }
+
+    fn get_hash_preimage(children: &ChildMap<F, ARITY>, hash: F) -> Option<&[F; ARITY]> {
+        children.get(&FWrap(hash))
+    }
+
+    // TODO: This is expensive and can be cached for the type.
+    fn init_empty(&mut self) {
+        self.empty_roots = [F::zero(); HEIGHT];
+
+        if HEIGHT == 0 {
+            return;
+        };
+
+        let mut preimage = [self.empty_roots[0]; ARITY];
+
+        for i in 0..HEIGHT {
+            let hash = self.register_hash(preimage);
+
+            self.empty_roots[i] = hash;
+
+            preimage = [hash; ARITY];
+            for elt in preimage.iter_mut().take(ARITY) {
+                *elt = hash;
+            }
+        }
+        self.count = 0;
+        self.root = self.empty_roots[HEIGHT - 1]
+    }
+
+    fn empty_root_for_height(&self, height: usize) -> F {
+        self.empty_roots[height - 1]
+    }
+
+    pub fn empty_root(&mut self) -> F {
+        self.empty_root_for_height(HEIGHT)
+    }
+    fn new(
```

Comment:
Don't we want to have one public constructor? Besides, an `impl Default` would be [nice to have](https://rust-lang.github.io/api-guidelines/interoperability.html#types-eagerly-implement-common-traits-c-common-traits)